hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/HoodieRecordBasedIndexingCatchup.java

```diff
@@ -0,0 +1,156 @@
+            case HoodieTimeline.COMMIT_ACTION:
+            case HoodieTimeline.DELTA_COMMIT_ACTION:
+            case HoodieTimeline.REPLACE_COMMIT_ACTION:
+              HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+              HoodieData<HoodieRecord> records = readRecordsFromDataFiles(instant);
+              metadataWriter.update(commitMetadata, records, instant.getTimestamp());
+              break;
+            case CLEAN_ACTION:
+              HoodieCleanMetadata cleanMetadata = CleanerUtils.getCleanerMetadata(metaClient, instant);
+              metadataWriter.update(cleanMetadata, instant.getTimestamp());
+              break;
+            case RESTORE_ACTION:
+              HoodieRestoreMetadata restoreMetadata = TimelineMetadataUtils.deserializeHoodieRestoreMetadata(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get());
+              metadataWriter.update(restoreMetadata, instant.getTimestamp());
+              break;
+            case ROLLBACK_ACTION:
+              HoodieRollbackMetadata rollbackMetadata = TimelineMetadataUtils.deserializeHoodieRollbackMetadata(
+                  metaClient.getActiveTimeline().getInstantDetails(instant).get());
+              metadataWriter.update(rollbackMetadata, instant.getTimestamp());
+              break;
+            default:
+              throw new IllegalStateException("Unexpected value: " + instant.getAction());
+          }
+        } catch (IOException e) {
+          throw new HoodieIndexException(String.format("Could not update metadata partition for instant: %s", instant), e);
+        } finally {
+          transactionManager.endTransaction(Option.of(instant));
+        }
+      }
+    }
+  }
+
+  private HoodieData<HoodieRecord> readRecordsFromDataFiles(HoodieInstant instant) throws IOException {
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+        .enable(true)
+        .build();
+    HoodieTableMetadata metadata = HoodieTableMetadata.create(engineContext, metadataConfig, metaClient.getBasePathV2().toString(), false);
+    HoodieMetadataFileSystemView fsView = new HoodieMetadataFileSystemView(metaClient, metaClient.getActiveTimeline().filter(i -> i.equals(instant)), metadata);
+    // Collect the list of latest base files present in each partition
+    List<String> partitions = metadata.getAllPartitionPaths();
+    fsView.loadAllPartitions();
+    final List<Pair<String, HoodieBaseFile>> partitionBaseFilePairs = new ArrayList<>();
+    for (String partition : partitions) {
+      partitionBaseFilePairs.addAll(fsView.getLatestBaseFiles(partition)
+          .map(basefile -> Pair.of(partition, basefile)).collect(Collectors.toList()));
+    }
+
+    return readRecordKeysFromBaseFiles(
+        engineContext,
+        partitionBaseFilePairs,
```

Comment:
should we be looking at base files alone? this sort of stuff lands us in trouble for e.g Flink will add inserts to logs

Reply:
We should be looking for file slices actually. I've changed it for the record index catchup task. I have a separate jira HUDI-6875 to remove the usages of `readRecordKeysFromBaseFiles` as it is being used at some other places too.